pycheribuild/projects/cross/newlib_rtems_riscv.py

```diff
@@ -0,0 +1,119 @@
+#
+# Copyright (c) 2019 Hesham Almatary
+# Copyright (c) 2017 Alex Richardson
+# All rights reserved.
+#
+# This software was developed by SRI International and the University of
+# Cambridge Computer Laboratory under DARPA/AFRL contract FA8750-10-C-0237
+# ("CTSRD"), as part of the DARPA CRASH research programme.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions
+# are met:
+# 1. Redistributions of source code must retain the above copyright
+#    notice, this list of conditions and the following disclaimer.
+# 2. Redistributions in binary form must reproduce the above copyright
+#    notice, this list of conditions and the following disclaimer in the
+#    documentation and/or other materials provided with the distribution.
+#
+# THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
+# ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+# ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
+# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+# OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+# HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+# OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+# SUCH DAMAGE.
+#
+from .crosscompileproject import *
+from ...utils import statusUpdate, IS_MAC, runCmd
+from ...config.loader import ComputedDefaultValue
+from pathlib import Path
+import tempfile
+
+
+class BuildNewlibRtemsRiscv(CrossCompileAutotoolsProject):
+    repository = GitRepository("https://github.com/CTSRD-CHERI/newlib")
+    projectName = "newlib-rtems-riscv"
+    make_kind = MakeCommandKind.GnuMake
+    rtems = True
+    add_host_target_build_config_options = False
+    # TODO Make this -O2 once a bug with LLVM/Clang relaxation is fixed
```

Comment:
I thought our believe was that it was because you compiled without `-mno-relax` and the linker used was LLD. Yes, `-mno-relax` currently messes up `%pcrel_lo` on master, but the fact that everything works at `-O0` is purely coincidence and has a 50% chance of failure (assuming this is a `.balign 4` that may or may not be aligned correctly pre-relaxation and gets put into `mtvec`.).

Reply:
> I think Alex's thinking is that any RISC-V-specific bits should go in BuildNewlibBaremetal, and any RTEMS-specific bits should go in BuildNewlibRtemsRiscv, which should be a subclass of BuildNewlibBaremetal.

RTEMS has its own GCC-newlib toolchain, so technically newlib-rtems isn't baremetal (and it doesn't use libgloss). That said, I am fine making it a subcalss of baremetal as it'll make things way easier for me